File: scripts/52_hogan_abnormal_day_counts.py
#!/usr/bin/env python3
"""Hogan 24 Aug 2026 alternative to Jingjing TV-day counts.

Leave-one-year-out day-of-year climatology from HKO dailyExtract 2012–2023.
Monthly counts of days above/below that climatology for tmean, tmax, tmin.
Also writes monthly mean RH and total rainfall as a check against climate_monthly.
REAL weather only. No health coefficients.
"""
from __future__ import annotations

import json
from collections import defaultdict
from datetime import date
from pathlib import Path
from statistics import mean

ROOT = Path("/workspace")
EXTRACT = ROOT / "data_raw/hko/daily_extract"
YEARS_CLIM = range(2012, 2024)
YEARS_STUDY = range(2013, 2024)


def parse_num(x):
    s = str(x).strip()
    if s in {"", "***", "----", "--", "NA", "N/A"}:
        return None
    if s.lower() == "trace":
        return 0.0
    try:
        return float(s)
    except ValueError:
        return None


def parse_year(path: Path, year: int) -> list[dict]:
    raw = json.loads(path.read_text())
    rows = []
    for mobj in raw["stn"]["data"]:
        mo = int(mobj["month"])
        for day_vec in mobj["dayData"]:
            label = str(day_vec[0]).strip()
            if not label.isdigit():
                continue
            day = int(label)
            rows.append(
                {
                    "date": date(year, mo, day),
                    "year": year,
                    "month": mo,
                    "day": day,
                    "month_day": f"{mo:02d}-{day:02d}",
                    # HKO dailyExtract order (0-based): day, pressure, tmax, tmean, tmin,
                    # dew point, RH%, cloud, rainfall mm, ...
                    "tmax": parse_num(day_vec[2]),
                    "tmean": parse_num(day_vec[3]),
                    "tmin": parse_num(day_vec[4]),
                    "rh": parse_num(day_vec[6]),
                    "rain": parse_num(day_vec[8]),
                }
            )
    return rows


def main() -> None:
    daily: list[dict] = []
    for y in YEARS_CLIM:
        path = EXTRACT / f"dailyExtract_{y}.xml"
        if not path.is_file():
            raise SystemExit(f"Missing {path}")
        daily.extend(parse_year(path, y))

    by_md_other: dict[tuple[int, str], dict[str, list[float]]] = defaultdict(
        lambda: {"tmean": [], "tmax": [], "tmin": []}
    )
    for row in daily:
        for y in YEARS_STUDY:
            if row["year"] == y:
                continue
            key = (y, row["month_day"])
            for k in ("tmean", "tmax", "tmin"):
                if row[k] is not None:
                    by_md_other[key][k].append(row[k])

    clim = {}
    for key, vals in by_md_other.items():
        clim[key] = {
            "tmean": mean(vals["tmean"]) if vals["tmean"] else None,
            "tmax": mean(vals["tmax"]) if vals["tmax"] else None,
            "tmin": mean(vals["tmin"]) if vals["tmin"] else None,
        }

    monthly: dict[str, dict] = {}
    for row in daily:
        if row["year"] not in YEARS_STUDY:
            continue
        mid = f"{row['year']:04d}-{row['month']:02d}"
        rec = monthly.setdefault(
            mid,
            {
                "month_id": mid,
                "year": row["year"],
                "month": row["month"],
                "n_days": 0,
                "n_days_tmean_above_doy": 0,
                "n_days_tmean_below_doy": 0,
                "n_days_tmax_above_doy": 0,
                "n_days_tmax_below_doy": 0,
                "n_days_tmin_above_doy": 0,
                "n_days_tmin_below_doy": 0,
                "rh": [],
                "rain": [],
            },
        )
        rec["n_days"] += 1
        c = clim.get((row["year"], row["month_day"]), {})
        for src, ckey, above, below in (
            ("tmean", "tmean", "n_days_tmean_above_doy", "n_days_tmean_below_doy"),
            ("tmax", "tmax", "n_days_tmax_above_doy", "n_days_tmax_below_doy"),
            ("tmin", "tmin", "n_days_tmin_above_doy", "n_days_tmin_below_doy"),
        ):
            obs, ref = row[src], c.get(ckey)
            if obs is None or ref is None:
                continue
            if obs > ref:
                rec[above] += 1
            elif obs < ref:
                rec[below] += 1
        if row["rh"] is not None:
            rec["rh"].append(row["rh"])
        if row["rain"] is not None:
            rec["rain"].append(row["rain"])

    rows_out = []
    for mid in sorted(monthly):
        rec = monthly[mid]
        rec["mean_relative_humidity"] = mean(rec["rh"]) if rec["rh"] else ""
        rec["total_rainfall_mm"] = sum(rec["rain"]) if rec["rain"] else ""
        del rec["rh"]
        del rec["rain"]
        rows_out.append(rec)

    if len(rows_out) != 132:
        raise SystemExit(f"expected 132 months, got {len(rows_out)}")

    cols = [
        "month_id",
        "year",
        "month",
        "n_days",
        "n_days_tmean_above_doy",
        "n_days_tmean_below_doy",
        "n_days_tmax_above_doy",
        "n_days_tmax_below_doy",
        "n_days_tmin_above_doy",
        "n_days_tmin_below_doy",
        "mean_relative_humidity",
        "total_rainfall_mm",
    ]
    text = ",".join(cols) + "\n"
    for rec in rows_out:
        text += ",".join(str(rec[c]) for c in cols) + "\n"

    processed = ROOT / "data_processed/hogan_abnormal_day_counts_2013_2023.csv"
    tables = ROOT / "outputs/tables/hogan_abnormal_day_counts_2013_2023.csv"
    tables.parent.mkdir(parents=True, exist_ok=True)
    processed.write_text(text)
    tables.write_text(text)

    jan = [r["n_days_tmean_below_doy"] for r in rows_out if r["month"] == 1]
    jul = [r["n_days_tmean_below_doy"] for r in rows_out if r["month"] == 7]
    meta = (
        "n_months,climatology,study_window,jan_mean_tmean_below,jul_mean_tmean_below,provenance\n"
        f"132,2012-2023 leave-one-year-out,2013-01 to 2023-12,"
        f"{sum(jan)/len(jan):.3f},{sum(jul)/len(jul):.3f},REAL_HKO_DAILY\n"
    )
    (ROOT / "outputs/tables/hogan_abnormal_day_counts_meta.csv").write_text(meta)
    print(f"wrote {processed} ({len(rows_out)} months)")
    print(f"Jan mean days tmean below DOY: {sum(jan)/len(jan):.1f}; Jul: {sum(jul)/len(jul):.1f}")


if __name__ == "__main__":
    main()

File: scripts/52_hogan_abnormal_day_counts.R
#!/usr/bin/env Rscript
# 52_hogan_abnormal_day_counts.R
# Hogan 24 Aug 2026 alternative to Jingjing TV-day counts:
# for each day of the year, take the historical mean of daily mean/max/min
# temperature; then count days in each calendar month that sit above or below
# that day-of-year average.
#
# Climatology: leave-one-year-out means of the same month-day across HKO
# dailyExtract years 2012–2023 (2024 held out of the study window).
# Ties with the climatology count as neither above nor below.
# REAL weather only. No health coefficients.

source(file.path("scripts", "utils.R"))
root <- project_root()
setwd(root)
cfg <- load_config(root)
ensure_packages(c("yaml", "jsonlite", "dplyr", "lubridate"))

extract_dir <- file.path(root, cfg$weather$sources$daily_extract_dir)
years_clim <- 2012:2023
years_study <- cfg$study$start_year:cfg$study$end_year

parse_year_extract <- function(path, year) {
  raw <- jsonlite::fromJSON(path, simplifyVector = FALSE)
  months <- raw$stn$data
  rows <- list()
  for (mobj in months) {
    mo <- as.integer(mobj$month)
    for (day_vec in mobj$dayData) {
      day_label <- trimws(as.character(day_vec[[1]]))
      if (!grepl("^[0-9]{1,2}$", day_label)) next
      day <- as.integer(day_label)
      rows[[length(rows) + 1]] <- data.frame(
        date = as.Date(sprintf("%04d-%02d-%02d", year, mo, day)),
        year = year,
        month = mo,
        day = day,
        tmax = parse_hko_number(day_vec[[3]]),
        tmean = parse_hko_number(day_vec[[4]]),
        tmin = parse_hko_number(day_vec[[5]]),
        relative_humidity = parse_hko_number(day_vec[[7]]),
        rainfall = parse_hko_number(day_vec[[9]]),
        stringsAsFactors = FALSE
      )
    }
  }
  dplyr::bind_rows(rows)
}

daily_list <- list()
for (y in years_clim) {
  path <- file.path(extract_dir, sprintf("dailyExtract_%d.xml", y))
  if (!file.exists(path)) {
    warning("Missing daily extract: ", path)
    next
  }
  daily_list[[as.character(y)]] <- parse_year_extract(path, y)
}
if (!length(daily_list)) stop("No dailyExtract files for climatology years")

weather_daily <- dplyr::bind_rows(daily_list) |>
  dplyr::arrange(date) |>
  dplyr::mutate(
    month_day = sprintf("%02d-%02d", month, day),
    month_id = format(date, "%Y-%m")
  )

# Leave-one-year-out climatology for each year in the study window.
study_daily <- weather_daily |>
  dplyr::filter(year %in% years_study)

clim_for_year <- function(y) {
  weather_daily |>
    dplyr::filter(year != y) |>
    dplyr::group_by(month_day) |>
    dplyr::summarise(
      clim_tmean = mean(tmean, na.rm = TRUE),
      clim_tmax = mean(tmax, na.rm = TRUE),
      clim_tmin = mean(tmin, na.rm = TRUE),
      n_clim_years = dplyr::n(),
      .groups = "drop"
    ) |>
    dplyr::mutate(year = y)
}

clim <- dplyr::bind_rows(lapply(years_study, clim_for_year))

tagged <- study_daily |>
  dplyr::left_join(clim, by = c("year", "month_day")) |>
  dplyr::mutate(
    tmean_above = as.integer(!is.na(tmean) & !is.na(clim_tmean) & tmean > clim_tmean),
    tmean_below = as.integer(!is.na(tmean) & !is.na(clim_tmean) & tmean < clim_tmean),
    tmax_above = as.integer(!is.na(tmax) & !is.na(clim_tmax) & tmax > clim_tmax),
    tmax_below = as.integer(!is.na(tmax) & !is.na(clim_tmax) & tmax < clim_tmax),
    tmin_above = as.integer(!is.na(tmin) & !is.na(clim_tmin) & tmin > clim_tmin),
    tmin_below = as.integer(!is.na(tmin) & !is.na(clim_tmin) & tmin < clim_tmin)
  )

monthly <- tagged |>
  dplyr::group_by(month_id, year, month) |>
  dplyr::summarise(
    n_days = dplyr::n(),
    n_days_tmean_above_doy = sum(tmean_above, na.rm = TRUE),
    n_days_tmean_below_doy = sum(tmean_below, na.rm = TRUE),
    n_days_tmax_above_doy = sum(tmax_above, na.rm = TRUE),
    n_days_tmax_below_doy = sum(tmax_below, na.rm = TRUE),
    n_days_tmin_above_doy = sum(tmin_above, na.rm = TRUE),
    n_days_tmin_below_doy = sum(tmin_below, na.rm = TRUE),
    mean_relative_humidity = mean(relative_humidity, na.rm = TRUE),
    total_rainfall_mm = sum(rainfall, na.rm = TRUE),
    .groups = "drop"
  ) |>
  dplyr::arrange(month_id)

stopifnot(nrow(monthly) == 132L)
stopifnot(all(monthly$n_days_tmean_above_doy + monthly$n_days_tmean_below_doy <= monthly$n_days))

out_processed <- file.path(root, "data_processed", "hogan_abnormal_day_counts_2013_2023.csv")
out_tables <- file.path(root, "outputs", "tables", "hogan_abnormal_day_counts_2013_2023.csv")
dir.create(dirname(out_tables), recursive = TRUE, showWarnings = FALSE)
utils::write.csv(monthly, out_processed, row.names = FALSE)
utils::write.csv(monthly, out_tables, row.names = FALSE)

meta <- data.frame(
  n_months = nrow(monthly),
  climatology_years = "2012-2023 leave-one-year-out",
  study_window = "2013-01 to 2023-12",
  mean_tmean_above = mean(monthly$n_days_tmean_above_doy),
  mean_tmean_below = mean(monthly$n_days_tmean_below_doy),
  jan_mean_tmean_below = mean(monthly$n_days_tmean_below_doy[monthly$month == 1L]),
  jul_mean_tmean_below = mean(monthly$n_days_tmean_below_doy[monthly$month == 7L]),
  provenance = "REAL_HKO_DAILY",
  stringsAsFactors = FALSE
)
utils::write.csv(
  meta,
  file.path(root, "outputs", "tables", "hogan_abnormal_day_counts_meta.csv"),
  row.names = FALSE
)

message("Wrote ", out_processed)
message(
  "Jan mean days tmean below DOY: ", round(meta$jan_mean_tmean_below, 1),
  "; Jul: ", round(meta$jul_mean_tmean_below, 1)
)
